assets/js/layouts/playground.js

```diff
@@ -1,69 +1,133 @@
 
-			if (link) {
+		for (const link of links) {
+			link.addEventListener("click", (e) => {
 				e.preventDefault();
 				this.switchTo(link, e.target);
-			}
-		});
-	}
+			});
+		}
+
+		// event listeners for theme toggles
+		const toggles = this.$el.querySelectorAll(
+			".playground-theme-toggle button"
+		);
+
+		for (const toggle of toggles) {
+			toggle.addEventListener("click", (e) => {
+				e.preventDefault();
 
-	get image() {
-		return this.$el.querySelector(".playground-header-figure-wrapper img");
+				// update theme
+				this.theme = e.target.closest("button").dataset.theme;
+
+				// reload current menu item with updated theme
+				const link = this.$el.querySelector(
+					".playground-header-menu a[aria-current]"
+				);
+				this.switchTo(link.href, link);
+			});
+		}
+	}
+	get figure() {
+		return this.$el.querySelector(".playground-header-figure");
 	}
 
 	async loadHtml(link) {
-		const response = await fetch(link);
+		const response = await fetch(link, {
+			cache: "no-store",
+		});
 		const body = await response.text();
 		const parser = new DOMParser();
 		const doc = parser.parseFromString(body, "text/html");
 
 		// load the new image as hidden before it gets injected
-		doc
-			.querySelector(".playground-header-figure-wrapper")
-			.classList.add("loading");
+		this.figure.classList.add("loading");
 
 		return doc;
 	}
 
-	preload(target) {
-		const newUrl = this.image.currentSrc.replace(
-			this.wrapper.dataset.image,
-			target.dataset.image
-		);
+	async replaceContent(link) {
+		const doc = await this.loadHtml(link);
+
+		this.$el.querySelector(".playground-backend").innerHTML = doc.querySelector(
+			".playground-backend"
+		).innerHTML;
+		this.$el.querySelector(".playground-filesystem").innerHTML =
+			doc.querySelector(".playground-filesystem").innerHTML;
+		this.$el.querySelector(".playground-medium").innerHTML =
+			doc.querySelector(".playground-medium").innerHTML;
+
+		// reactivate code highlighting
+		Prism.highlightAll();
+	}
+
+	async replaceImage(target) {
+		return new Promise((resolve) => {
+			const theme = this.theme.charAt(0).toUpperCase() + this.theme.slice(1);
+			const oldImage = this.wrapper.querySelector("img:last-child");
+
+			// clone image, replace src and srcset
+			// and add it to the wrapper
+			const newImage = oldImage.cloneNode(true);
+			newImage.src = target.dataset[`image${theme}Src`];
+			newImage.srcset = target.dataset[`image${theme}Srcset`];
+
+			// when new image is loaded…
+			newImage.onload = async () => {
+				// wait briefly to prevent flickering
+				await new Promise((resolve) => {
+					setTimeout(() => {
+						resolve();
+					}, 50);
+				});
+
+				// fade out old image
+				oldImage.style.opacity = 0;
+
+				// remove old image after fade out is complete
+				setTimeout(() => {
+					oldImage.remove();
+				}, 700);
 
-		new Image().src = newUrl;
+				resolve();
+			};
+
+			// append new image behind current image
+			this.wrapper.appendChild(newImage);
```

Comment:
`Finding`: I think the flickering on quick theme or story changes happens because multiple new images get loaded into the DOM before the first old one is removed.

`Idea` how we could solve this:
- When a `newImage` is appended to the DOM, set a global flag like `pendingUpdate`.
- Before executing the logic in `replaceImage()`, check if the flag is set. If so, set the new target to another global prop like `finalTarget` and return immediately.
- Before resolving the top-level promise, unset the `pendingUpdate` prop. Also check if something is in the `finalTarget` prop. If so, unset it and call `replaceImage(finalTarget)` to start a second update cycle for the final target.

Maybe it could be done even cleaner with a throttle helper, but I think this should already work quite well.